Order List redesign - Step 1: Added TabLayout to the order list

anitaa1990 opened the pull request:
Fixes #1399. This PR completes the first step of the Stats Improvements Project as defined in the master thread [here](https://github.com/woocommerce/woocommerce-android/issues/1398). 

### Changes
- Added a new class `OrderListView` that handles the order list related logic such as populating the `RecyclerView`, displaying empty view.
- Added `TabLayout` with `Processing` and `All Orders` tab and anchor the `TabLayout` to the top of the view when scrolled.
- Added logic to remember the last selected tab and default to that when order list screen is loaded.

### Notes
- The filter option will be removed in a separate PR when addressing #1400. 
- The search revamp will take place in a separate PR when addressing #1400.  
- Requires a design review from @kyleaparker before PR can be merged.
- **Note that this PR is to be merged to a feature branch.**

### Testing
- Run `OrderListItemTest`
- Test marking an order as complete from the `Processing` tab.
- Test changing an order status from the `All Orders` tab.
- Test pagination working as expected for both tabs.
- Search should work as it is currently.

### Screenshots
<img width="300" src="https://user-images.githubusercontent.com/22608780/64146200-a64c8080-ce39-11e9-9eb8-8a022d59ba1a.png"> . <img width="300" src="https://user-images.githubusercontent.com/22608780/64146202-a8aeda80-ce39-11e9-85df-a92d41b84970.png">

#### Empty View
<img width="300" src="https://user-images.githubusercontent.com/22608780/64146226-c11ef500-ce39-11e9-9f9f-cf5c057b6070.png">

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

nbradbury commented:
@anitaa1990 The tabs look great! A few comments before I review the code:

* I'm thinking the empty view when Processing is selected should say something like "No processing orders," since the user may have other orders
* I noticed that the pull-to-refresh spinner shows every time you switch tabs. Would it make more sense to only show it when the user actually does a PTR, or when the list is empty?
* When returning to the order list after marking an order as completed, the PTR spinner shows and the completed order remains in the Processing list until the network request completes. Since the status will have been changed in the local DB, could we simply refresh the list from the DB when returning to the list from the detail?
* If I switch tabs while a network request is still processing, the list is updated with the wrong orders. Example below: I switched to All Orders while processing orders were still being fetched, and when the request completed only processing orders show. I think there needs to be some sort of check that the completed request is for the correct tab. 

![Screenshot_1567519624](https://user-images.githubusercontent.com/3903757/64180567-faa03200-ce32-11e9-8cd9-c11b90bb177b.png)

anitaa1990 commented:
Thanks for the quick review @nbradbury! 

 > I'm thinking the empty view when Processing is selected should say something like "No processing orders," since the user may have other orders.

As Kylea mentioned, I modified it to `No orders to process` in d85c66e

> When returning to the order list after marking an order as completed, the PTR spinner shows and the completed order remains in the Processing list until the network request completes. Since the status will have been changed in the local DB, could we simply refresh the list from the DB when returning to the list from the detail?

This looks to be the current behaviour in the app at the moment. Debugging this further, I found that the local db is not updated with the status change until after the network request completes. So the UI only updates after the status change is successfully completed.

> I noticed that the pull-to-refresh spinner shows every time you switch tabs. Would it make more sense to only show it when the user actually does a PTR, or when the list is empty?

I was looking into how to implement this and noticed that we are deleting the local db every time a new order list is fetched from db. So the first time, when `Processing` tab is clicked, we are updating the db with just `processing` orders [here](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/83770f7e7a46f526ac8759cf551dac9b2a031c98/plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCOrderStore.kt#L492). Then once the `All orders` tab is clicked, the db is updated with all order statuses. If we initiate PTR again on `Processing` tab, the order list is updated with just processing orders and the other orders are deleted so we are forced to refresh again. I'm not sure if we can fix this on the app side and would probably need a change in FluxC. I'll open a separate issue/PR in FluxC for this.

> If I switch tabs while a network request is still processing, the list is updated with the wrong orders. Example below: I switched to All Orders while processing orders were still being fetched, and when the request completed only processing orders show. I think there needs to be some sort of check that the completed request is for the correct tab.

I've added logic to update the order list only if the order filter status matches the current tab in fc86b31 but I am still able to reproduce this because of the issue mentioned above. Would you mind if we handle this scenario separately?

nbradbury commented:
> I'm not sure if we can fix this on the app side and would probably need a change in FluxC. I'll open a separate issue/PR in FluxC for this.

Ah, that all makes sense. I think we can hold off fixing this in FluxC for now. We've been talking about a strategy for purging older/outdated data and I'm thinking this could be part of that rather than handled separately. Make sense?

> Would you mind if we handle this scenario separately?

Sure, that's fine.

Please give me a ping when this passes design review and I'll do a more thorough code review.

kyleaparker commented:
Can we go with the simplified empty state for now until the design is ready for that screen:

<img width="317" alt="Stats Improvements 2019-09-05 09-11-16" src="https://user-images.githubusercontent.com/712873/64344818-38c95d00-cfbd-11e9-97a4-39bb06650985.png">

The "Share your store" CTA doesn't seem like the best option for this screen. Everything else looks good.

anitaa1990 commented:
Thanks @kyleaparker for the review!  This is how it looks now:
<img width="300" src="https://user-images.githubusercontent.com/22608780/64358797-bd02fb00-d024-11e9-9404-320f10e26b19.png"> . <img width="300" src="https://user-images.githubusercontent.com/22608780/64358802-beccbe80-d024-11e9-9d41-063d13d605c8.png">

anitaa1990 commented:
> Ah, that all makes sense. I think we can hold off fixing this in FluxC for now. We've been talking about a strategy for purging older/outdated data and I'm thinking this could be part of that rather than handled separately. Make sense?

Yup, sounds good! I think this is already fixed in Amanda's [Order list optimisation PR](https://github.com/woocommerce/woocommerce-android/pull/1289) 👍

anitaa1990 commented:
Update: @kyleaparker has provided some temp screens for handling the empty view. I've opened a separate issue [here](https://github.com/woocommerce/woocommerce-android/issues/1411) to tackle this since this PR already exceeds the maximum code limit. 

@nbradbury , since I will be tackling the empty state handling in a separate PR, we can go ahead with the review for this one 👍

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/OrderListFragment.kt:
```diff
@@ -228,47 +227,63 @@ class OrderListFragment : TopLevelFragment(), OrderListContract.View,
+                    tab_layout.addTab(tab)
 
-        presenter.takeView(this)
+                    // Start with the given time period selected
```
Minor nit, and I won't hold up merging due to this, but this comment is confusing because it doesn't involve a "time period."

anitaa1990 replied:
nice catch! Fixed in 3750e5a

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/OrderListFragment.kt:
```diff
@@ -228,47 +227,63 @@ class OrderListFragment : TopLevelFragment(), OrderListContract.View,
+
+            override fun onTabUnselected(tab: TabLayout.Tab) {}
+
+            override fun onTabReselected(tab: TabLayout.Tab) {}
```
I think the general behavior when a tab is re-selected is to scroll to the top, as we do with the bottom nav tabs.

anitaa1990 replied:
👍 Fixed in 3750e5a

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/OrderListFragment.kt:
```diff
@@ -228,47 +227,63 @@ class OrderListFragment : TopLevelFragment(), OrderListContract.View,
+            override fun onTabReselected(tab: TabLayout.Tab) {}
+        })
+
+        // set activity toolbar elevation to 0
```
Sorry to keep being so nitpicky, but this comment isn't very helpful. It's already clear from the code that the elevation is being set to 0, so the comment should instead state _why_ it's being set to 0. More importantly, shouldn't the elevation be restored when this tab is no longer active?

anitaa1990 replied:
Good point! Fixed in 69bc34a

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/OrderListFragment.kt:
```diff
@@ -314,47 +329,37 @@ class OrderListFragment : TopLevelFragment(), OrderListContract.View,
     override fun showOrders(orders: List<WCOrderModel>, filterByStatus: String?, isFreshData: Boolean) {
-        orderStatusFilter = filterByStatus
+        // Only update the order list view if the new filter match the currently selected order status
+        if (orderStatusFilter == filterByStatus) {
```
👍

nbradbury commented on WooCommerce/src/main/res/layout/order_list_view.xml:
```diff
@@ -0,0 +1,46 @@
+        android:focusable="true">
+
+        <!-- Orders List View -->
+        <LinearLayout
```
Is this `LinearLayout` necessary?

anitaa1990 replied:
Good point here! I don't think this is needed. Fixed in 33ca592

nbradbury commented:
@anitaa1990 I'm done with my first round. Everything looks good apart from my usual assortment of nitpicks :)

anitaa1990 commented:
Thanks for the quick review @nbradbury! Fixed all the nitpicks and PR ready for another round 👍
